### PR TITLE
CC-3158: Improved Logging

### PR DESCRIPTION
**Pull request by cchristiezen**
Adding additional logging for backup and restore process so the user knows it is still running.  Also, moved logging over to logrus where appropriate.

### CODE REVIEW

File: dfs/backup.go [hunk 1]
```diff
@@ -35,64 +35,92 @@ const (
...
-			glog.Errorf("Could not find image %s: %s", image, err)
+			plog.WithFields(log.Fields{"image": image}).WithError(err).
+				Error("Could not find image for backup")
 			return err
 		}
-		glog.Infof("Prepared Docker image %s for backup", image)
+
+		plog.WithFields(log.Fields{
+			"image":          image,
+			"numberComplete": i + 1,
```
Comment (smousa) [inline dfs/backup.go]: field names must be all lower case

File: dfs/backup.go [hunk 2]
```diff
@@ -183,17 +228,17 @@ func (dfs *DistributedFilesystem) writeBackupMetadata(data BackupInfo, w *tar.Wr
 		jsonData []byte
 		err      error
 	)
-	glog.V(2).Infof("Writing backup metadata")
+	plog.Info("Writing backup metadata")
```
Comment (smousa) [inline dfs/backup.go]: Should this be debug level?

File: dfs/backup.go [hunk 3]
```diff
@@ -183,17 +228,17 @@ func (dfs *DistributedFilesystem) writeBackupMetadata(data BackupInfo, w *tar.Wr
...
 	if jsonData, err = json.Marshal(data); err != nil {
 		return err
 	}
 	header := &tar.Header{Name: BackupMetadataFile, Size: int64(len(jsonData))}
 	if err := w.WriteHeader(header); err != nil {
-		glog.V(2).Infof("Could not create metadata header for backup: %s", err)
+		plog.WithError(err).Info("Could not create metadata header for backup")
```
Comment (smousa) [inline dfs/backup.go]: Should this be debug level?

File: dfs/backup.go [hunk 4]
```diff
@@ -183,17 +228,17 @@ func (dfs *DistributedFilesystem) writeBackupMetadata(data BackupInfo, w *tar.Wr
...
 		return err
 	}
 	if _, err := w.Write(jsonData); err != nil {
-		glog.V(2).Infof("Could not write backup metadata: %s", err)
+		plog.WithError(err).Info("Could not write backup metadata")
```
Comment (smousa) [inline dfs/backup.go]: Should this be debug level?

File: dfs/docker/docker.go [hunk 5]
```diff
@@ -93,8 +98,26 @@ func (d *DockerClient) SaveImages(images []string, writer io.Writer) error {
 		Names:        images,
 		OutputStream: writer,
 	}
-	glog.Infof("Exporting images %s", images)
-	return d.dc.ExportImages(opts)
+
+	startTime := time.Now()
+	plog.Info("Starting image export", images)
```
Comment (smousa) [inline dfs/docker/docker.go]: This log message looks wrong--do you need WithField?

File: dfs/backup.go [hunk 6]
```diff
@@ -102,26 +130,40 @@ func (dfs *DistributedFilesystem) Backup(data BackupInfo, w io.Writer) error {
...
+
 	// dump the images from all the snapshots into the backup
 	imageReader, errchan := dfs.dockerSavePipe(images...)
+
+	plog.Info("Starting export of images to backup")
 	if err := rewriteTar(DockerImagesFile, tarOut, imageReader); err != nil {
 		// be a good citizen and clean up any running threads
 		<-errchan
-		glog.Errorf("Could not write images %v to backup: %s", images, err)
+		plog.WithFields(log.Fields{"images": images}).
```
Comment (smousa) [inline dfs/backup.go]: Use WithField instead of WithFields when you are adding a single field

File: dfs/backup.go [hunk 7]
```diff
@@ -102,26 +130,40 @@ func (dfs *DistributedFilesystem) Backup(data BackupInfo, w io.Writer) error {
...
+			WithError(err).
+			Error("Could not write images to backup")
 		return err
 	} else if err := <-errchan; err != nil {
-		glog.Errorf("Could not export images %v for backup: %s", images, err)
+		plog.WithFields(log.Fields{"images": images}).
```
Comment (smousa) [inline dfs/backup.go]: Use WithField instead of WithFields when you are adding a single field

File: dfs/progress.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+)
+
+type ProgressCounter struct {
+	Total                 uint64
+	Message               string
+	lastUpdate            time.Time
+	updateIntervalSeconds int
+}
+
+func (pc *ProgressCounter) Write(data []byte) (int, error) {
```
Comment (smousa) [inline dfs/progress.go]: Tests?

File: dfs/restore.go [hunk 9]
```diff
@@ -31,7 +32,7 @@ var (
 
 // Restore restores application data from a backup.
 func (dfs *DistributedFilesystem) Restore(r io.Reader, version int) error {
-	glog.Infof("Detected backup version %d", version)
+	plog.Infof(fmt.Sprintf("Detected backup version %d", version))
```
Comment (smousa) [inline dfs/restore.go]: Nope, gotta use WithField here

File: dfs/restore.go [hunk 10]
```diff
@@ -197,24 +202,26 @@ func (dfs *DistributedFilesystem) restoreV1(r io.Reader) error {
...
+				plog.Info("Loading docker images from backup")
 				writer, errc := dfs.imageLoadPipe()
 				tarwriter := tar.NewWriter(writer)
 				s = &stream{tarwriter: tarwriter, writer: writer, errc: errc}
 				streamMap[DockerImagesFile] = s
 			}
 			hdr.Name = parts[1]
 			if err := s.tarwriter.WriteHeader(hdr); err != nil {
-				glog.Errorf("Could not write image header %s: %s", hdr.Name, err)
+				plog.WithError(err).WithFields(log.Fields{"header": hdr.Name}).
```
Comment (smousa) [inline dfs/restore.go]: Use WithField when you are adding a single field

File: dfs/restore.go [hunk 11]
```diff
@@ -197,24 +202,26 @@ func (dfs *DistributedFilesystem) restoreV1(r io.Reader) error {
...
+					Error("Could not write image header")
 				dataError = err
 				return err
 			} else if _, err := io.Copy(s.tarwriter, backuptar); err != nil {
-				glog.Errorf("Could not write image data with header %s: %s", hdr.Name, err)
+				plog.WithError(err).WithFields(log.Fields{"header": hdr.Name}).
```
Comment (smousa) [inline dfs/restore.go]: Same

File: dfs/restore_test.go [hunk 12]
```diff
@@ -54,7 +54,7 @@ func (s *DFSTestSuite) TestRestore_LoadImages(c *C) {
 	err := tarfile.WriteHeader(&tar.Header{Name: DockerImagesFile, Size: 0})
 	c.Assert(err, IsNil)
 	tarfile.Close()
-	s.docker.On("LoadImage", mock.AnythingOfType("*tar.Reader")).Return(nil)
+	s.docker.On("LoadImage", mock.Anything).Return(nil)
```
Comment (smousa) [inline dfs/restore_test.go]: Why did you change the type in these tests?

**Comment by smousa** [comment]
Here is an example of how to do logging: https://github.com/control-center/serviced/blob/develop/container/endpoint.go

File: dfs/restore_test.go (hunk 12 above)
Comment (cchristiezen) [inline dfs/restore_test.go]: The reader was wrapped with a TeeReader to so I could count the bytes.  The TeeReader type (io.teeReader) is not exported so I couldn't change it to that.  I tried the interface io.Reader, but that didn't work either with AnythingOfType.  I changed it to use Anything instead.  When its used, the value gets casted in the next line which would through a panic if something was wrong, so I think it will still catch issues that appear.

File: dfs/backup.go (hunk 1 above)
Comment (cchristiezen) [inline dfs/backup.go]: Changed

File: dfs/backup.go (hunk 6 above)
Comment (cchristiezen) [inline dfs/backup.go]: Changed

File: dfs/backup.go (hunk 7 above)
Comment (cchristiezen) [inline dfs/backup.go]: Changed

File: dfs/backup.go (hunk 2 above)
Comment (cchristiezen) [inline dfs/backup.go]: Changed

File: dfs/backup.go (hunk 3 above)
Comment (cchristiezen) [inline dfs/backup.go]: Changed

File: dfs/backup.go (hunk 4 above)
Comment (cchristiezen) [inline dfs/backup.go]: Changed

File: dfs/progress.go (hunk 8 above)
Comment (cchristiezen) [inline dfs/progress.go]: Added

File: dfs/restore.go (hunk 9 above)
Comment (cchristiezen) [inline dfs/restore.go]: Changed

File: dfs/restore.go (hunk 10 above)
Comment (cchristiezen) [inline dfs/restore.go]: Changed

File: dfs/restore.go (hunk 11 above)
Comment (cchristiezen) [inline dfs/restore.go]: Changed

**Comment by cchristiezen** [comment]
retest this

File: dfs/progress.go [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+package dfs
+
+import (
+	"time"
+
+	"github.com/control-center/serviced/logging"
+)
+
+var (
+	plog = logging.PackageLogger()
```
Comment (smousa) [inline dfs/progress.go]: This should be in dfs.go I think, since this is used across the whole package

File: dfs/docker/docker.go [hunk 14]
```diff
@@ -93,8 +97,26 @@ func (d *DockerClient) SaveImages(images []string, writer io.Writer) error {
 		Names:        images,
 		OutputStream: writer,
 	}
-	glog.Infof("Exporting images %s", images)
-	return d.dc.ExportImages(opts)
+
+	startTime := time.Now()
+
+	plog.WithField("images", images).Info("Starting image export")
```
Comment (cchristiezen) [inline dfs/docker/docker.go]: At this point there isn't anything more to add than the list of images and logging it with every message might be too much.

File: dfs/backup.go [hunk 15]
```diff
@@ -35,94 +35,137 @@ const (
 // Backup writes all application data into an export stream
 func (dfs *DistributedFilesystem) Backup(data BackupInfo, w io.Writer) error {
```
Comment (cchristiezen) [inline dfs/backup.go]: This was added.
